smarts/core/utils/traffic_history_serivce.py

```diff
@@ -0,0 +1,159 @@
+        """prepare 1 batch ahead, when received request, immediately return the previously
+        prepared batch and prepares the next batch.
+        """
+        return_batch = {}
+        while True:
+            historyRange = request_queue.get()
+            assert isinstance(historyRange, RequestHistoryRange)
+            send_data_conn.send(return_batch)
+            return_batch = {}
+            with open(history_file_path, "rb") as f:
+                for index, (t, vehicles_state) in enumerate(
+                    ijson.kvitems(f, "", use_float=True)
+                ):
+                    if (
+                        historyRange.start_index <= index
+                        and index < historyRange.start_index + historyRange.batch_count
+                    ):
+                        return_batch[t] = vehicles_state
+        send_data_conn.close()
+
+    @property
+    def all_timesteps(self):
+        return self._all_timesteps
+
+    @property
+    def history_file_path(self):
+        return self._history_file_path
+
+    @property
+    def traffic_history(self):
+        return {**self._current_traffic_history, **self._prev_batch_history}
+
+    def _prepare_next_batch(self):
+        self._request_queue.put(
+            RequestHistoryRange(
+                start_index=self._range_start,
+                batch_count=self._batch_size,
+            )
+        )
+        self._range_start += self._batch_size
+
+    def fetch_history_at_timestep(self, timestep: str):
+        if timestep not in self._all_timesteps:
+            return {}
+        elif timestep in self.traffic_history:
+            return self.traffic_history[timestep]
+
+        # ask child process to prepare the next batch:
+        self._prepare_next_batch()
+        self._prev_batch_history = self._current_traffic_history
+        # receives the previous batch child process prepared
+        self._current_traffic_history = self._receive_data_conn.recv()
+        if timestep in self._current_traffic_history:
+            return self._current_traffic_history[timestep]
+        # no history exists at requested timestamp
+        return {}
+
+    @staticmethod
+    def fetch_agent_missions(history_file_path):
+        vehicle_missions = {}
+        with open(history_file_path, "rb") as f:
+            for t, vehicles_state in ijson.kvitems(f, "", use_float=True):
+                for vehicle_id in vehicles_state:
+                    if vehicle_id in vehicle_missions:
+                        continue
+                    vehicle_missions[vehicle_id] = scenario.Mission(
+                        start=scenario.Start(
+                            vehicles_state[vehicle_id]["position"][:2],
+                            scenario.Heading(vehicles_state[vehicle_id]["heading"]),
+                        ),
+                        goal=scenario.EndlessGoal(),
+                        start_time=float(t),
+                    )
```

Comment:
Are the iterations of the `for` loop independent, or can be made independent, such that they can be replaced with `map` or other constructs to improve performance and speed?